test/saml.spec.js

```diff
@@ -163,3 +164,124 @@ describe("signon#completeSamlLogin", () => {
+
+  it(`should make request to token endpoint then redirect browser to SAML logout
+      endpoint`, async () => {
+    // Create mock response for token request
+    const mockTokenResponse = {
+      data: { token: "foo-bar-1234" },
+    };
+    axios.get.mockImplementationOnce(() => mockTokenResponse);
+
+    // Access token, ID token, and user should all exist before and after logging
+    // out of service provider (should not log user out of tenant application, only
+    // the service provider)
+    expect(Cookies.get(`access.${tenantId}`)).toBeTruthy();
+    expect(Cookies.get(`id.${tenantId}`)).toBeTruthy();
+    expect(Userfront.tokens.accessToken).toBeTruthy();
+    expect(Userfront.tokens.idToken).toBeTruthy();
+    expect(Userfront.user.userId).toEqual(33);
+    expect(Userfront.user.email).toEqual("johndoe@example.com");
+
+    const accessToken = Userfront.tokens.accessToken;
+    await logout({ method: "saml" });
+
+    // Assert GET /auth/saml/idp/token request was made
+    expect(axios.get).toHaveBeenCalledWith(
+      `${Userfront.store.baseUrl}auth/saml/idp/token`,
+      {
+        headers: {
+          authorization: `Bearer ${accessToken}`,
+        },
+      }
+    );
+
+    // Assert client was redirected to /auth/saml/idp/logout with token
+    expect(window.location.assign).toHaveBeenCalledWith(
+      `${Userfront.store.baseUrl}auth/saml/idp/logout` +
+        `?tenant_id=${tenantId}` +
+        `&token=${mockTokenResponse.data.token}` +
+        `&uuid=${Userfront.user.userUuid}`
+    );
+
+    // Should not have cleared the access and ID tokens
+    expect(Cookies.get(`access.${tenantId}`)).toBeTruthy();
+    expect(Cookies.get(`id.${tenantId}`)).toBeTruthy();
+    expect(Userfront.tokens.accessToken).toBeTruthy();
+    expect(Userfront.tokens.idToken).toBeTruthy();
+
+    // Should not have cleared the user object
+    expect(Userfront.user.userId).toBeTruthy();
+    expect(Userfront.user.email).toBeTruthy();
+    expect(Userfront.user.update).toBeTruthy();
+  });
+
+  it(`should return early if store.tokens.accessToken isn't defined`, async () => {
+    // Init without access token
+    Cookies.set(`access.${tenantId}`, "", {});
+    Cookies.set(`id.${tenantId}`, mockIdToken, {});
+    Userfront.init(tenantId);
+    expect(Userfront.tokens.idToken).toBeTruthy();
+    expect(Userfront.user.userId).toEqual(33);
+    expect(Userfront.user.email).toEqual("johndoe@example.com");
+
+    await logout({ method: "saml" });
+
+    // Should not have made request to /auth/saml/idp/token or redirected the user
+    expect(axios.get).not.toHaveBeenCalled();
+    expect(window.location.assign).not.toHaveBeenCalled();
+
+    // Should not have modified ID token or user
+    expect(Cookies.get(`id.${tenantId}`)).toBeTruthy();
+    expect(Userfront.tokens.idToken).toBeTruthy();
+    expect(Userfront.user.userId).toBeTruthy();
+    expect(Userfront.user.email).toBeTruthy();
+    expect(Userfront.user.update).toBeTruthy();
+  });
```

Comment:
Does this mean that if the user initiates a logout on their SP and is not logged into Userfront already, nothing will happen? E.g. they will be stuck on the Userfront login screen, and then logging in will not do anything related to SAML?

Reply:
If they log in, they will be brought back to the Login page and this would be run again.

I suppose it would be helpful to return an error rather than nothing, but also sounds awkward (e.g. `Please log in to complete SAML logout process.`). Will update accordingly.